### PR TITLE
 Add DB bg update to cleanup extremities.

### PR DESCRIPTION
**Pull request by erikjohnston**
Based on #5274.

Due to #5269 we may have extremities in our DB that we shouldn't have, so lets add a cleanup task such to remove those.

### CODE REVIEW

**Comment by erikjohnston** [comment]
Hmm, this isn't happy on jki.re.

File: synapse/storage/events.py [hunk 1]
```diff
@@ -2341,6 +2347,188 @@ def get_all_updated_current_state_deltas_txn(txn):
...
+            to_delete = set()
+            while non_rejected_leaves:
+                event_id = non_rejected_leaves.pop()
+                prev_event_ids = graph.get(event_id, set())
+                non_rejected_leaves.update(prev_event_ids)
+                to_delete.update(prev_event_ids)
+
+            to_delete.intersection_update(original_set)
+
+            logger.info("Deleting up to %d forward extremities", len(to_delete))
```
Comment (richvdh) [inline synapse/storage/events.py]: this looks a bit redundant?

File: synapse/storage/events.py [hunk 2]
```diff
@@ -2341,6 +2347,188 @@ def get_all_updated_current_state_deltas_txn(txn):
...
+
+            deleted = self._simple_delete_many_txn(
+                txn=txn,
+                table="event_forward_extremities",
+                column="event_id",
+                iterable=to_delete,
+                keyvalues={},
+            )
+
+            logger.info("Deleted %d forward extremities", deleted)
```
Comment (richvdh) [inline synapse/storage/events.py]: might be handy to log the size of the `original_set` here too. Also some context on why this is happening:

```suggestion
            logger.info("Deleted %d forward extremities of %d checked, to clean up #5269", deleted, len(original_set))
```

File: synapse/storage/events.py [hunk 3]
```diff
@@ -2341,6 +2347,188 @@ def get_all_updated_current_state_deltas_txn(txn):
...
+                LEFT JOIN events USING (event_id)
+                LEFT JOIN event_json USING (event_id)
+                LEFT JOIN rejections USING (event_id)
+                """, (batch_size,)
+            )
+
+            for prev_event_id, event_id, metadata, rejected, outlier in txn:
+                original_set.add(prev_event_id)
+
+                if not event_id or outlier:
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: `_calculate_new_extremities` does check they're not outliers fwiw, which is the main reason I did it.

File: synapse/storage/events.py [hunk 4]
```diff
@@ -2341,6 +2347,188 @@ def get_all_updated_current_state_deltas_txn(txn):
...
+                    ",".join("?" for _ in to_check),
+                )
+                txn.execute(sql, to_check)
+
+                for prev_event_id, event_id, metadata, rejected in txn:
+                    if event_id in graph:
+                        # Already handled this event previously, but we still
+                        # want to record the edge.
+                        graph.setdefault(event_id, set()).add(prev_event_id)
+                        logger.info("Already handled")
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: Woops, that was debug logging

File: synapse/storage/events.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/events.py]: oh, right

File: synapse/storage/events.py [hunk 5]
```diff
@@ -221,6 +221,7 @@ class EventsStore(
 ):
     EVENT_ORIGIN_SERVER_TS_NAME = "event_origin_server_ts"
     EVENT_FIELDS_SENDER_URL_UPDATE_NAME = "event_fields_sender_url"
+    EVENT_DELETE_SOFT_FAILED_EXTREMITIES = "delete_soft_failed_extremities"
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: I think the idea is to namespace them since they're part of a mixin

[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/events.py]: hrm. I thought the idea was to more-or-less match the name of the update in the database.

[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: That makes more sense. Will change it to that.

File: synapse/storage/events.py [hunk 6]
```diff
@@ -2341,6 +2347,188 @@ def get_all_updated_current_state_deltas_txn(txn):
...
+            txn.execute(
+                """SELECT prev_event_id, event_id, internal_metadata,
+                    rejections.event_id IS NOT NULL, events.outlier
+                FROM (
+                    SELECT event_id AS prev_event_id
+                    FROM _extremities_to_check
+                    LIMIT ?
+                ) AS f
+                LEFT JOIN event_edges USING (prev_event_id)
+                LEFT JOIN events USING (event_id)
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: From #synapse-dev: let's leave for now, hopefully the fix to the comment makes it less confusing

File: synapse/storage/events_bg_updates.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+            # from the forward extremities table.
+            non_rejected_leaves = set()
+
+            # Set of event IDs that have been soft failed, and for which we
+            # should check if they have descendants which haven't been soft
+            # failed.
+            soft_failed_events_to_lookup = set()
+
+            # First, we get `batch_size` events from the table, pulling out
+            # their successor events, if any, and their successor events
```
Comment (richvdh) [inline synapse/storage/events_bg_updates.py]:
```suggestion
            # their successor events, if any, and the successor events'
```

**Pull request merged**